File: src/Core/Models/ReadmeData.cs
namespace CyberPatriotAutomation.Core.Models;

/// <summary>
/// Represents parsed data from a CyberPatriot README file
/// </summary>
public class ReadmeData
{
    /// <summary>
    /// Title of the competition round
    /// </summary>
    public string Title { get; set; } = string.Empty;

    /// <summary>
    /// Operating system specified in the README
    /// </summary>
    public string OperatingSystem { get; set; } = string.Empty;

    /// <summary>
    /// Competition scenario description
    /// </summary>
    public string Scenario { get; set; } = string.Empty;

    /// <summary>
    /// List of authorized administrators with their passwords
    /// </summary>
    public List<AuthorizedUser> Administrators { get; set; } = new();

    /// <summary>
    /// List of authorized standard users
    /// </summary>
    public List<AuthorizedUser> Users { get; set; } = new();

    /// <summary>
    /// Software that should be installed/updated
    /// </summary>
    public List<SoftwareRequirement> RequiredSoftware { get; set; } = new();

    /// <summary>
    /// Software that should be removed (prohibited)
    /// </summary>
    public List<string> ProhibitedSoftware { get; set; } = new();

    /// <summary>
    /// Services that must be running
    /// </summary>
    public List<string> CriticalServices { get; set; } = new();

    /// <summary>
    /// Services that should not be running
    /// </summary>
    public List<string> ProhibitedServices { get; set; } = new();

    /// <summary>
    /// Groups that need to be created with their members
    /// </summary>
    public List<GroupRequirement> GroupRequirements { get; set; } = new();

    /// <summary>
    /// New user accounts that need to be created
    /// </summary>
    public List<string> UsersToCreate { get; set; } = new();

    /// <summary>
    /// Competition guidelines and warnings
    /// </summary>
    public List<string> Guidelines { get; set; } = new();

    /// <summary>
    /// Actionable items extracted from paragraph tags
    /// </summary>
    public List<ActionableItem> ActionableItems { get; set; } = new();

    /// <summary>
    /// Raw sections extracted from the README
    /// </summary>
    public Dictionary<string, string> Sections { get; set; } = new();
}

/// <summary>
/// Represents an actionable item extracted from the README
/// </summary>
public class ActionableItem
{
    public ActionableItemType Type { get; set; }
    public string Description { get; set; } = string.Empty;
    public string RawText { get; set; } = string.Empty;
    public Dictionary<string, string> Details { get; set; } = new();
}

/// <summary>
/// Types of actionable items that can be extracted
/// </summary>
public enum ActionableItemType
{
    CreateUser,
    CreateGroup,
    AddUserToGroup,
    RemoveUserFromGroup,
    EnableService,
    DisableService,
    InstallSoftware,
    RemoveSoftware,
    ConfigureSetting,
    SecurityPolicy,
    FileOperation,
    Other,
}

/// <summary>
/// Represents an authorized user from the README
/// </summary>
public class AuthorizedUser
{
    public string Username { get; set; } = string.Empty;
    public string? Password { get; set; }
    public bool IsAdmin { get; set; }
    public bool IsPrimaryUser { get; set; }
    public string? Notes { get; set; }
}

/// <summary>
/// Represents a software requirement from the README
/// </summary>
public class SoftwareRequirement
{
    public string Name { get; set; } = string.Empty;
    public string? Version { get; set; }
    public bool ShouldBeLatest { get; set; }
    public bool IsRequired { get; set; } = true;
    public string? Notes { get; set; }
}

/// <summary>
/// Represents a group that needs to be created
/// </summary>
public class GroupRequirement
{
    public string GroupName { get; set; } = string.Empty;
    public List<string> Members { get; set; } = new();
}

File: rust/src/models/readme_data.rs
use std::collections::HashMap;

/// Represents parsed data from a CyberPatriot README file.
#[derive(Debug, Default, Clone)]
pub struct ReadmeData {
    pub title: String,
    pub operating_system: String,
    pub scenario: String,
    pub administrators: Vec<AuthorizedUser>,
    pub users: Vec<AuthorizedUser>,
    pub required_software: Vec<SoftwareRequirement>,
    pub prohibited_software: Vec<String>,
    pub critical_services: Vec<String>,
    pub prohibited_services: Vec<String>,
    pub group_requirements: Vec<GroupRequirement>,
    pub users_to_create: Vec<String>,
    pub guidelines: Vec<String>,
    pub actionable_items: Vec<ActionableItem>,
    /// Raw sections extracted from the README (header -> content). Case-insensitive lookup helper below.
    pub sections: HashMap<String, String>,
}

impl ReadmeData {
    /// Case-insensitive section lookup, mirroring the C# OrdinalIgnoreCase dictionary.
    pub fn section(&self, header: &str) -> Option<&String> {
        self.sections
            .iter()
            .find(|(k, _)| k.eq_ignore_ascii_case(header))
            .map(|(_, v)| v)
    }
}

/// Represents an actionable item extracted from the README.
#[derive(Debug, Default, Clone)]
pub struct ActionableItem {
    pub item_type: ActionableItemType,
    pub description: String,
    pub raw_text: String,
    pub details: HashMap<String, String>,
}

/// Types of actionable items that can be extracted.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default)]
pub enum ActionableItemType {
    CreateUser,
    CreateGroup,
    AddUserToGroup,
    RemoveUserFromGroup,
    EnableService,
    DisableService,
    InstallSoftware,
    RemoveSoftware,
    ConfigureSetting,
    SecurityPolicy,
    FileOperation,
    #[default]
    Other,
}

/// Represents an authorized user from the README.
#[derive(Debug, Default, Clone)]
pub struct AuthorizedUser {
    pub username: String,
    pub password: Option<String>,
    pub is_admin: bool,
    pub is_primary_user: bool,
    pub notes: Option<String>,
}

/// Represents a software requirement from the README.
#[derive(Debug, Clone)]
pub struct SoftwareRequirement {
    pub name: String,
    pub version: Option<String>,
    pub should_be_latest: bool,
    pub is_required: bool,
    pub notes: Option<String>,
}

impl Default for SoftwareRequirement {
    fn default() -> Self {
        Self {
            name: String::new(),
            version: None,
            should_be_latest: false,
            is_required: true,
            notes: None,
        }
    }
}

/// Represents a group that needs to be created.
#[derive(Debug, Default, Clone)]
pub struct GroupRequirement {
    pub group_name: String,
    pub members: Vec<String>,
}
